capsules/extra/src/bmi270.rs

```diff
@@ -0,0 +1,529 @@
+                        self.state.set(State::Sleep);
+                        self.buffer.replace(buffer);
+                        self.ninedof_client
+                            .map(|client| client.callback(i2c_err as usize, 0, 0));
+                    } else {
+                        self.state.set(State::InitWriteConfig);
+                    }
+                } else if us == ALARM_TIME_20000 {
+                    buffer[0] = Registers::InternalStatus as u8;
+
+                    if let Err((i2c_err, buffer)) = self.i2c.write_read(buffer, 1, 1) {
+                        self.state.set(State::Sleep);
+                        self.buffer.replace(buffer);
+                        self.ninedof_client
+                            .map(|client| client.callback(i2c_err as usize, 0, 0));
+                    } else {
+                        self.state.set(State::CheckStatus);
+                    }
+                }
+            }
+            _ => {}
+        });
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> NineDof<'a> for BMI270<'a, A, I> {
+    fn set_client(&self, client: &'a dyn NineDofClient) {
+        self.ninedof_client.set(client);
+    }
+
+    fn read_accelerometer(&self) -> Result<(), ErrorCode> {
+        if !self.pending_accel.get() {
+            self.pending_accel.set(true);
+            if self.pending_gyro.get() {
+                Ok(())
+            } else {
+                self.start_measurement()
+            }
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    fn read_gyroscope(&self) -> Result<(), ErrorCode> {
+        if !self.pending_gyro.get() {
+            self.pending_gyro.set(true);
+            if self.pending_accel.get() {
+                Ok(())
+            } else {
+                self.start_measurement()
+            }
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> AlarmClient for BMI270<'a, A, I> {
+    fn alarm(&self) {
+        self.handle_alarm()
+    }
+}
+
+impl<'a, A: Alarm<'a>, I: I2CDevice> I2CClient for BMI270<'a, A, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if let Err(i2c_err) = status {
+            match self.state.get() {
+                State::InitWriteConfig => {
+                    self.config_file.replace(buffer);
+                }
+                _ => {
+                    self.buffer.replace(buffer);
+                }
+            }
+            self.state.set(State::Sleep);
+            self.ninedof_client
+                .map(|client| client.callback(i2c_err as usize, 0, 0));
+            return;
+        }
+
+        match self.state.get() {
+            State::WaitingForAlarm(us) => {
+                self.buffer.replace(buffer);
+                let delay = self.alarm.ticks_from_us(us);
+                self.alarm.set_alarm(self.alarm.now(), delay);
+            }
+            State::InitWriteConfig => {
```

Comment:
I believe the problem is that here, you don't replace the buffer into `self.buffer` so that when you finally reach the `InitDone` state, `self.buffer` is empty and you never enter the closer, so don't send any further I2C commands and the driver is just stuck.

Reply:
Replacing the buffer seemed to fix that bug, although I'm still getting 0's when I read the sensor (both for accelerometer and gyroscope).

Reply:
OK, so it's "working" now, but the readings are totally wrong (they are zero). That's progress!